compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -4625,56 +4662,150 @@ private STNode parseObjectConstructorExpression(STNode annots) {
 
     /**
      * <p>
-     * Parse object type qualifiers.
+     * Parse object constructor qualifiers.
      * </p>
-     * <code>object-type-qualifiers := [client]</code>
+     * <code>object-constructor-qualifier := [client]</code>
      *
      * @return Parsed node
      */
-    private STNode parseObjectTypeQualifiers() {
+    private STNode parseObjectConstructorQualifiers() {
+        STToken nextToken = peek();
+        STToken firstQualifier;
+        switch (nextToken.kind) {
+            case CLIENT_KEYWORD:
+                firstQualifier = consume();
+                break;
+            case READONLY_KEYWORD:
+            case ABSTRACT_KEYWORD:
+            case ISOLATED_KEYWORD:
+                // Here we allow parsing of old object type qualifiers (`abstract` and `readonly`)
+                // and then log an error
+                firstQualifier = consume();
+                addInvalidNodeToNextToken(firstQualifier, DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED,
+                        firstQualifier.text());
+                return parseObjectConstructorQualifiers();
+            case OBJECT_KEYWORD:
+                return STNodeFactory.createEmptyNodeList();
+            default:
+                recover(nextToken, ParserRuleContext.OBJECT_CONSTRUCTOR_QUALIFIER);
+                return parseObjectConstructorQualifiers();
+        }
+
+        return parseObjectConstructorNextQualifiers(firstQualifier);
+    }
+
+    private STNode parseObjectConstructorNextQualifiers(STNode firstQualifier) {
```

Comment:
Do we need two methods?
I think the one with the loop would be sufficient to handle all cases.